app/scripts/metamask-controller.js

```diff
@@ -5322,27 +5326,207 @@ export default class MetamaskController extends EventEmitter {
   }
 
   /**
-   * Requests approval for permissions for the specified origin
+   * Prompts the user with permittedChains approval for given chainId.
    *
-   * @param origin - The origin to request approval for.
-   * @param permissions - The permissions to request approval for.
+   * @param {string} origin - The origin to request approval for.
+   * @param {Hex} chainId - The chainId to add incrementally.
    */
-  async requestPermissionApprovalForOrigin(origin, permissions) {
+  async requestApprovalPermittedChainsPermission(origin, chainId) {
     const id = nanoid();
-    return this.approvalController.addAndShowApprovalRequest({
+    await this.approvalController.addAndShowApprovalRequest({
       id,
       origin,
       requestData: {
         metadata: {
           id,
           origin,
         },
-        permissions,
+        permissions: {
+          [PermissionNames.permittedChains]: {
+            caveats: [
+              {
+                type: CaveatTypes.restrictNetworkSwitching,
+                value: [chainId],
+              },
+            ],
+          },
+        },
       },
       type: MethodNames.RequestPermissions,
     });
   }
 
+  /**
+   * Requests permittedChains permission for the specified origin
+   * and replaces any existing CAIP-25 permission with a new one.
+   *
+   * @param {string} origin - The origin to request approval for.
+   * @param {Hex} chainId - The chainId to add incrementally.
+   * @param {boolean} autoApprove - If the chain should be granted without prompting for user approval.
```

Comment:
Should we include an explanation of why we might want to auto approve in some cases: i.e. currently chainPermissions may be granted as part of a `wallet_addEthereumChain` request in which the permissions and addition of the network are combined into one confirmation

Reply:
https://github.com/MetaMask/metamask-extension/pull/29213/commits/e741f74abb2b602da2aa52a77915c89c48c3c3f8